### ISSUE TITLE
Core dump on addition backprop

### ISSUE THREAD

**Issue by azimi99** [azimi99/micronn#1]
There's a bug on section of code:
src/tensor.cpp
```
    if (out.requires_grad) {
        
        out._prev = {this, const_cast<Tensor*>(&other)};

        out._backward = [this, other_ptr = const_cast<Tensor*>(&other), out]() {
            for (size_t i = 0; i < data.size(); ++i) {
                std::cout << "grad[" << i << "] = " << out.grad[i] << std::endl;
                if (requires_grad) grad[i] += out.grad[i];
                if (other_ptr->requires_grad) other_ptr->grad[i] += out.grad[i];
            }
        };
    }
```

I'm not sure why this is happening but expect to set the gradients to 1.0 in main.cpp to verify it's working correctly. Also when done add the unit test in `tests/`

```
TEST_CASE("Test addition operation and backprop", "[tensor]"){

    Tensor a({1.0, 2.0, 3.0, 4.0}, {4});
    Tensor b({1.0, 2.0, 3.0, 4.0}, {4});
    Tensor t = a + b;
    t.backward();
    REQUIRE(t.shape[0] == 4);
    // TODO complete test case
}
```
